### PR TITLE
Escape name section ids to be WebAssembly spec compatible.

### PR DESCRIPTION
**Pull request by yurydelendik**
wasm-dis does not properly generate id for the functions, e.g. `$core::ptr::drop_in_place::ha8f9534d58f8a1fd (.llvm.3277592684094598524)` will be enclosed in the quotes instead of being escaped.

The wasm2asm fails to proceed by replacing `(call "$core::ptr::drop_in_place::ha8f9534d58f8a1fd (.llvm.3277592684094598524)" ..` with `(call 0...` (due to the autodrop pass)

TODO:
- [x] testing
- [ ] <del>(optional) remove https://github.com/WebAssembly/binaryen/blob/master/src/passes/Print.cpp#L38-L40></del>
- [ ] <del>(optional) fail when `(call "something"` is found</del>

### CODE REVIEW

**Comment by kripken** [comment]
I'm not sure about the example - is it not ok to simply quote that string, which seems to have no `"` or `\`s? Is the issue that right now we don't quote or parse quotes in the call instructions to a function with a weird name? That is, it seems like in your example, we could write/read this text:

```
(func "$core::ptr::drop_in_place::ha8f9534d58f8a1fd (.llvm.3277592684094598524)"
  ..
)
(func $caller-of-that
  (call "$core::ptr::drop_in_place::ha8f9534d58f8a1fd (.llvm.3277592684094598524)")
)
```

If so, then this seems more like an s-expression parsing/emitting bug than binary format? However, it seems like `(func "..")` has a meaning actually, to export it, so perhaps that's not good?

**Comment by kripken** [comment]
Talking on IRC, the approach here seems necessary as `(call "quoted")` is not valid wat, so sounds good.

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -565,6 +565,35 @@ void WasmBinaryWriter::writeInlineString(const char* name) {
   }
 }
 
+inline bool isHexDigit(char ch) {
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: these could be `static` (no need for `inline` I don't think)

File: test/complexNames.wasm.fromBinary [hunk 2]
```diff
@@ -0,0 +1,11 @@
+(module
+ (type $0 (func))
+ (export "$foo\20\28.bar\29" (func $1))
+ (func $foo\20\28.bar\29 (; 0 ;) (type $0)
+  (nop)
+ )
+ (func $1 (; 1 ;) (type $0)
```
Comment (kripken) [inline test/complexNames.wasm.fromBinary]: does the binary not have a names section name for this function?

Comment (yurydelendik) [reply test/complexNames.wasm.fromBinary]: it has -- wasm-as exported name as "1"

File: src/wasm/wasm-binary.cpp (hunk 1 above)
Comment (yurydelendik) [reply src/wasm/wasm-binary.cpp]: fixed

File: src/passes/Print.cpp [hunk 3]
```diff
@@ -35,12 +35,7 @@ static bool isFullForced() {
 }
 
 static std::ostream& printName(Name name, std::ostream& o) {
-  // we need to quote names if they have tricky chars
```
Comment (kripken) [inline src/passes/Print.cpp]: What about the case where the name has not arrived from the binary reading code? I think we may still need to escape there.

**Comment by binji** [comment]
FYI: there was some discussion about a similar topic here: https://github.com/WebAssembly/spec/issues/617. Andrew suggested that we should be able to do what you described here, something like `(call $"foo")`. Andreas suggested instead that we provide an annotation to create the correct name section info: `(call $foo (@name "foo"))`. Anyway, if you have opinions about either of these, feel free to mention it on that issue.

**Comment by yurydelendik** [comment]
> annotation to create the correct name section info: (call $foo (@name "foo"))

Binaryen is using names section as source of the identifiers. Ideally, it is a one-way conversion: name section id can be missing or non-unique, so binaryen mutates the original name to fit its needs. Addressing WebAssembly/spec#617 issue might not fully cover the issue we are having with using name section data as function/local names. Annotations will be a way to go.

Currently, this patch is only fixing `wasm-dis` output when weird names are present in the name section (and somewhat recover the name section when `wasm-as -g` is used)

**Comment by kripken** [comment]
Makes sense, this PR lgtm, and we can see about other escaping issues in the text format etc. later.

File: src/passes/Print.cpp (hunk 3 above)
Comment (kripken) [reply src/passes/Print.cpp]: based on irc discussion, I think we still need quoting here - we need to emit something we can parse.

we should probably check if we need to quote, as escaping may have removed the problem for us. if not, then we must quote.

File: test/complexNames.wast.from-wast [hunk 4]
```diff
@@ -0,0 +1,10 @@
+(module
+ (type $0 (func))
+ (export "$foo\20\28.bar\29" (func $1))
+ (func $foo\20\28.bar\29 (; 0 ;) (type $0)
+  (nop)
+ )
+ (func $1 (; 1 ;) (type $0)
```
Comment (kripken) [inline test/complexNames.wast.from-wast]: interesting that the name ended up `$1` here - I guess that's a limitation of the s-expression parsing?

File: test/complexNames.wasm.fromBinary [hunk 5]
```diff
@@ -0,0 +1,11 @@
+(module
+ (type $0 (func))
+ (export "$foo\20\28.bar\29" (func $1))
```
Comment (kripken) [inline test/complexNames.wasm.fromBinary]: Hmm, are we escaping both the export and the function name identically? If so then we would be regressing compared to before, as reading and writing a binary won't preserve export names?

File: test/complexNames.wast.from-wast (hunk 4 above)
Comment (yurydelendik) [reply test/complexNames.wast.from-wast]: I think it is default name that is coming from https://github.com/WebAssembly/binaryen/blob/master/src/wasm/wasm-binary.cpp#L1034

File: test/complexNames.wasm.fromBinary (hunk 5 above)
Comment (yurydelendik) [reply test/complexNames.wasm.fromBinary]: It just create this name from wast I typed in the complexNames.wast below -- I can retype it unescaped.

File: test/complexNames.wast.from-wast (hunk 4 above)
Comment (kripken) [reply test/complexNames.wast.from-wast]: Got it, makes sense.

File: test/complexNames.wasm.fromBinary (hunk 5 above)
Comment (kripken) [reply test/complexNames.wasm.fromBinary]: I thought this file was the result of `wasm-dis tests/complexNames.wasm`, that is, printing out the binary you added to the test suite?

If you added both a wast and a wasm, I might be better to give them different names and not just differentiate them by the suffix - can be confusing.

Comment (yurydelendik) [reply test/complexNames.wasm.fromBinary]: fixed

**Comment by kripken** [comment]
Thanks, looks like the test outputs need to be updated now though.

Also please rename `tests/complexNames.wasm => tests/complexBinaryNames.wasm` and `tests/complexNames.wast => tests/complexTextNames.wast`, that way it will be clearer which test output is which, as I was trying to say before (sorry if I wasn't clear).

**Comment by kripken** [comment]
Perfect, thanks @yurydelendik! Last thing, sorry, we had some breakage on master - it's fixed now, can you please merge it in here, so we can see that tests pass?

**Comment by kripken** [comment]
Thanks @yurydelendik!

**Pull request merged**